version: "3"

services:
  namenode:
    image: hadoop-namenode
    container_name: namenode
    restart: always
    volumes:
      - namenode:/opt/volume/namenode
    ports:
      - "9870:9870"
  datanode:
    image: hadoop-datanode
    container_name: datanode
    restart: always
    depends_on:
      - namenode
    volumes:
      - datanode:/opt/volume/datanode
  resourcemanager:
    image: hadoop-resourcemanager
    container_name: resourcemanager
    restart: always
    depends_on:
      - namenode
    ports:
      - "8088:8088"
  nodemanager:
    image: hadoop-nodemanager
    container_name: nodemanager
    restart: always
    depends_on:
      - namenode
    ports:
      - "8042:8042"
  historyserver:
    image: hadoop-historyserver
    container_name: historyserver
    restart: always
    volumes:
      - historyserver:/hadoop/yarn/timeline
    depends_on:
      - namenode
  hive-server:
    image: hadoop-hive
    container_name: hiveserver
    command:
      - /usr/local/bin/script_hive.sh
      - bash -c 'if [[ $$(schematool -dbType mysql -info | grep "*** schemaTool failed ***") ]]; then schematool -dbType mysql -initSchema; fi'
      - hive --service
    restart: always
    depends_on:
      - namenode
    ports:
      - "10000:10000"
  mysql-server:
    image: mysql:5.7-debian
    command:
      --default-authentication-plugin=mysql_native_password
    restart: always
    volumes:
      - mysql-server:/var/lib/mysql
    environment:
      MYSQL_ROOT_PASSWORD: mysql
    ports:
      - "33306:3306"
  jupyter:
    image: hadoop-jupyter
    container_name: jupyter
    command: /opt/anaconda/bin/jupyter notebook --ip='*' --port=8888 --notebook-dir=/opt/notebooks --no-browser --allow-root
    restart: always
    volumes:
      - jupyter:/opt/notebooks
    ports:
      - "8888:8888"

volumes:
  namenode:
  datanode:
  historyserver:
  mysql-server:
  jupyter:
